calnex/firmware/ossfirmware.go

```diff
@@ -31,12 +32,34 @@ type OSSFW struct {
 
 // NewOSSFW returns initialized version of OSSFW
 func NewOSSFW(source string) (*OSSFW, error) {
+	var err error
+	var vs string
+	var v *version.Version
+
 	fw := &OSSFW{
 		filepath: source,
 	}
 	basename := filepath.Base(fw.filepath)
-	vs := strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentinel_fw_v", "")
-	v, err := version.NewVersion(strings.ToLower(vs))
+	// Extract version from filename
+	// sentinel_fw_v2.13.1.0.5583D-20210924.tar
+	// calnex_combined_fw_R21.0.0.9705-20241111.tar
+	if strings.HasPrefix(basename, "sentinel_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentinel_fw_", "")
+	} else if strings.HasPrefix(basename, "sentry_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "sentry_fw_", "")
+	} else if strings.HasPrefix(basename, "calnex_combined_fw_") {
+		vs = strings.ReplaceAll(strings.TrimSuffix(basename, filepath.Ext(basename)), "calnex_combined_fw_", "")
+	} else {
+		return nil, fmt.Errorf("Unexpected file string, expected sentinel_fw_ sentry_fw_ or calnex_combined_fw_  at start: %s", basename)
+	}
+
+	if strings.HasPrefix(vs, "v") {
+		v, err = version.NewVersion(strings.SplitN(strings.ToLower(vs), ".", 2)[1])
+	} else if strings.HasPrefix(vs, "R") {
+		v, err = version.NewVersion(strings.TrimPrefix(strings.ToLower(vs), "r"))
+	} else {
+		return nil, fmt.Errorf("Unexpected file string, expected v2. or R. : %s", vs)
+	}
```

Comment:
Can you please cover this beautiful logic with unit tests so we never break it by mistake?